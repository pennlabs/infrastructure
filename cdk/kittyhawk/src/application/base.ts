import { Construct } from "constructs";
import { removeSubdomain } from "..";
import { Certificate } from "../certificate";
import { Deployment, DeploymentProps } from "../deployment";
import { Ingress, IngressProps } from "../ingress";
import { Service } from "../service";
import { ServiceAccount } from "../serviceaccount";

export interface ApplicationProps {
  /**
   * Ingress configuration
   */
  readonly ingress?: IngressProps;

  /**
   * Deployment configuration
   */
  readonly deployment: DeploymentProps;

  /**
   * Port to expose the application on.
   */
  readonly port?: number;

  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || "undefined_release";
    const fullname = `${release_name}-${appname}`;

    new Service(this, fullname, props.port);

    if (props.createServiceAccount) {
      new ServiceAccount(this, `${appname}-${release_name}`, {
        serviceAccountName: release_name,
      });
    }

    new Deployment(this, fullname, {
      ...props.deployment,
      port: props.port,
      ...(props.createServiceAccount
        ? { serviceAccountName: release_name }
        : {}),
    });

    if (props.ingress) {
      new Ingress(this, fullname, props.ingress);

      const alreadyCreatedCertificates = new Set<string>();

      props.ingress.rules.forEach((rules) => {
        const finalDomain: string = removeSubdomain(
          rules.host,
          rules.isSubdomain ?? false
        );
        if (!alreadyCreatedCertificates.has(finalDomain)) {
          new Certificate(this, fullname, rules);
          alreadyCreatedCertificates.add(finalDomain);
        }
      });
    }
  }
}
