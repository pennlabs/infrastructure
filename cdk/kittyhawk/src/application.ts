import { Construct } from 'constructs';
import { Deployment, DeploymentProps } from './deployment';
import { Ingress, IngressProps } from './ingress';
import { Service, ServiceProps } from './service'
import { Certificate } from './certificate';

/**
 * Warning: Before editing any interfaces, make sure that none of the interfaces will have 
 * property names that conflict with each other. Typescript may not throw an error and it 
 * could cause problems.
 */
export interface ApplicationProps extends IngressProps, DeploymentProps,
  ServiceProps { }


export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${appname}`

    new Service(this, fullname, props);

    new Deployment(this, fullname, props);

    if (props.ingress) {
      new Ingress(this, fullname, props)

      new Certificate(this, fullname, props)
    }
  }
}

/**
 * Helper function that first checks to make sure that the environment variable array 
 * doesn't already contain the env var, then inserts it into the array.
 * @param envArray array of environment variables to insert into 
 * @param envKey name of the environment variable to insert
 * @param envValue value of the environment variable
 */
function insertIfNotPresent(envArray: { name: string, value: string }[], envKey: string, envValue: any) {
  const envSettingsModule = envArray?.filter(env => (env.name === envKey));
  if (envSettingsModule?.length > 0) {
    throw new Error(`${envKey} should not be redefined as an enviroment variable.`)
  }
  envArray.push({ name: envKey, value: envValue });

}
export interface DjangoApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * DJANGO_SETTINGS_MODULE environment variable.
   */
  readonly djangoSettingsModule: string;
}

export class DjangoApplication extends Application {
  constructor(scope: Construct, appname: string, props: DjangoApplicationProps) {

    // Have to be careful here with references when mutating things
    let djangoExtraEnv = Array.from(props.extraEnv || []);

    // Insert DJANGO_SETTINGS_MODULE and DOMAIN
    insertIfNotPresent(djangoExtraEnv, 'DJANGO_SETTINGS_MODULE', props.djangoSettingsModule)
    insertIfNotPresent(djangoExtraEnv, 'DOMAIN', props.domain)

    // Configure the ingress using ingressPaths.
    const djangoIngress = [{ host: props.domain, paths: props.ingressPaths }]

    // If everything passes, construct the Application.
    super(scope, appname, {
      ...props,
      extraEnv: djangoExtraEnv,
      ingress: djangoIngress,
    });
  }
}

export interface ReactApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * PORT environment variable for react. Default '80'.
   */
  readonly portEnv?: string;
}

export class ReactApplication extends Application {
  constructor(scope: Construct, appname: string, props: ReactApplicationProps) {

    // Have to be careful here with references when mutating things
    let reactExtraEnv = Array.from(props.extraEnv || []);

    // Insert DOMAIN and PORT as env vars.
    insertIfNotPresent(reactExtraEnv, 'DOMAIN', props.domain)
    insertIfNotPresent(reactExtraEnv, 'PORT', props.portEnv || '80')

    // Configure the ingress using ingressPaths.
    const reactIngress = [{ host: props.domain, paths: props.ingressPaths }]

    // If everything passes, construct the Application.
    super(scope, appname, {
      ...props,
      extraEnv: reactExtraEnv,
      ingress: reactIngress,
    });
  }
}

export class RedisApplication extends Application {
  constructor(scope: Construct, appname: string, redisProps: Partial<ApplicationProps>) {
    super(scope, appname, {
      ...redisProps,
      image: 'redis',
      tag: redisProps.tag || '6.0',
      port: redisProps.port || 6379,
    });
  }
}
