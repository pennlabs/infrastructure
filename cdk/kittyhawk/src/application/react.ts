import { Construct } from "constructs";
import { DeploymentProps } from "../deployment";
import { HostRules, IngressProps } from "../ingress";
import { Application } from "./base";

export interface ReactApplicationProps {
  /**
   * Deployment configuration
   */
  readonly deployment: DeploymentProps;

  /**
   * Port to expose the application on.
   */
  readonly port?: number;

  /**
   * Host is the domain the application runs on,
   * isSubdomain is true if the domain should be treated as a subdomain for certificate purposes.
   * paths is the list of paths to expose the application on.
   * See the certificate documentation for more details.
   *
   */
  readonly domain: HostRules;

  /**
   * Optional ingressProps to override the default ingress props.
   */
  readonly ingressProps?: IngressProps;

  /**
   * PORT environment variable for react. Default '80'.
   */
  readonly portEnv?: string;

  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class ReactApplication extends Application {
  constructor(scope: Construct, appname: string, props: ReactApplicationProps) {
    // Now, we ensure there are no duplicate env variables, even if they redefine it
    const reactExtraEnv = [
      ...new Set([
        ...(props.deployment?.env || []),
        { name: "DOMAIN", value: props.domain.host },
        { name: "PORT", value: props.portEnv || "80" },
      ]),
    ];

    // If everything passes, construct the Application.
    super(scope, appname, {
      port: props.port ?? 80,
      deployment: {
        ...props.deployment,
        env: reactExtraEnv,
      },
      ingress: {
        rules: [props.domain],
        ...props.ingressProps,
      },
      createServiceAccount: props.createServiceAccount,
    });
  }
}
