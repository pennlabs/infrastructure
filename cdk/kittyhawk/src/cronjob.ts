import { Construct } from 'constructs';
import { KubeCronJobV1Beta1 as CronJobApiObject } from '../imports/k8s';
import { Container, ContainerProps, Volume } from './container';

export interface CronJobProps extends ContainerProps {

  /**
     * The schedule in Cron format.
     */
  readonly schedule: string;

  /**
    * Restart policy for all containers.
    * 
    * @default "Never"
    * 
    */

  readonly restartPolicy?: 'Always' | 'OnFailure' | 'Never';

  /**
    * The number of successful finished jobs to retain. 
    * 
    * @default 1
    * 
    */
  readonly successLimit?: number;

  /**
    * The number of failed jobs to retain.
    * 
    * @default 1
    * 
    */
  readonly failureLimit?: number;

  /**
    * Secret volume mounts for cronjob container.
    *
    * @default undefined
    */
  readonly secretMounts?: { name: string, mountPath: string, subPath: string }[]

}


export class CronJob extends Construct {
  constructor(scope: Construct, jobname: string, props: CronJobProps) {
    super(scope, jobname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${jobname}`
    const containers: Container[] = [new Container(props)];
    const volumes: Volume[] | undefined = props.secretMounts?.map(m => new Volume(m))

    new CronJobApiObject(this, `cronjob-${fullname}`, {
      metadata: {
        name: fullname,
        namespace: 'default',
        labels: { name: fullname },
      },
      spec: {
        schedule: props.schedule,
        jobTemplate: {
          spec: {
            template: {
              spec: {
                restartPolicy: props.restartPolicy || 'Never',
                containers: containers,
                volumes: volumes,
              },
            },

          },
        },
        failedJobsHistoryLimit: props.failureLimit ?? 1,
        successfulJobsHistoryLimit: props.successLimit ?? 1,
      },
    });
  }
}