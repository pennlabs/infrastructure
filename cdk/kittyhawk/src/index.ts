
export * from './application';
export * from './autoscaler';
export * from './certificate';
export * from './chart';
export * from './container';
export * from './cronjob';
export * from './deployment';
export * from './ingress';
export * from './service';
