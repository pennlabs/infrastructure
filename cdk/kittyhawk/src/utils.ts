export type NonEmptyArray<T> = [T, ...T[]];

export function nonEmptyMap<T, U>(
  arr: NonEmptyArray<T>,
  callbackfn: (value: T) => U
): NonEmptyArray<U> {
  return arr.map(callbackfn) as any as NonEmptyArray<U>;
}
