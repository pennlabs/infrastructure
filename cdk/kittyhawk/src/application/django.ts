import { Construct } from 'constructs';
import { Application } from '.';
import { DeploymentProps } from '../deployment';
import { HostRules, IngressProps } from '../ingress';
import { NonEmptyArray, nonEmptyMap } from '../utils';

export interface DjangoApplicationProps {
  /**
   * Deployment configuration
   */
  readonly deployment: DeploymentProps;

  /**
   * Port to expose the application on.
   */
  readonly port?: number;

  /**
   * Array of domain(s). Host is the domain the application runs on,
   * and isSubdomain is true if the domain should be treated as a subdomain for certificate purposes.
   * See the certificate documentation for more details.
   */
  readonly domains: NonEmptyArray<{ host: string; isSubdomain?: boolean }>;

  /**
   * Just the list of paths passed to the ingress since we already know the host. Optional.
   *
   * @default undefined
   */
  readonly ingressPaths?: string[];

  /**
   * Optional ingressProps to override the default ingress props.
   */
  readonly ingressProps?: IngressProps;

  /**
   * DJANGO_SETTINGS_MODULE environment variable.
   */
  readonly djangoSettingsModule: string;

  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}


export default class DjangoApplication extends Application {
  constructor(scope: Construct, appname: string, props: DjangoApplicationProps) {

    // Now, we ensure there are no duplicate env variables, even if they redefine it
    const djangoExtraEnv = [...new Set([
      ...props.deployment?.env || [],
      { name: 'DJANGO_SETTINGS_MODULE', value: props.djangoSettingsModule },
      { name: 'DOMAIN', value: nonEmptyMap(props.domains, (h => h.host)).join() },
    ])];

    // Configure the ingress using ingressPaths if ingressPaths is defined.
    const djangoIngress: HostRules[] = props.domains?.map(h => ({
      host: h.host,
      paths: props.ingressPaths || [],
      isSubdomain: h.isSubdomain ?? false,
    }));

    // If everything passes, construct the Application.
    super(scope, appname, {
      port: props.port ?? 80,
      deployment: {
        ...props.deployment,
        env: djangoExtraEnv,
      },
      ingress: { 
        rules: djangoIngress as NonEmptyArray<HostRules>,
        ...props.ingressProps,
       },
      createServiceAccount: props.createServiceAccount,
    });
  }
}