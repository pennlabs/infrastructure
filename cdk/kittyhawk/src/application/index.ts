import { Construct } from 'constructs';
import { Certificate } from '../certificate';
import { Deployment, DeploymentProps } from '../deployment';
import { Ingress, IngressProps } from '../ingress';
import { Service } from '../service';
import { ServiceAccount } from '../serviceaccount';
import { nonEmptyMap } from '../utils';

import DjangoApplication from './django';
import ReactApplication from './react';
import RedisApplication from './redis';

export { DjangoApplication };
export { ReactApplication };
export { RedisApplication };

/**
 * Warning: Before editing any interfaces, make sure that none of the interfaces will have
 * property names that conflict with each other. Typescript may not throw an error and it
 * could cause problems.
 */
export interface ApplicationProps {
  readonly ingress?: IngressProps;
  readonly deployment: DeploymentProps;
  readonly port?: number;
  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${appname}`;

    new Service(this, fullname, props.port);

    if (props.createServiceAccount) {
      new ServiceAccount(this, `${appname}-${release_name}`, {
        serviceAccountName: release_name
      })
    }

    new Deployment(this, fullname, {
      ...props.deployment,
      ...(props.createServiceAccount ? { serviceAccountName: release_name } : {}),
    });

    if (props.ingress) {
      new Ingress(this, fullname, props.ingress);

      nonEmptyMap(props.ingress.rules, rule => {
        return new Certificate(this, fullname, rule);
      });
    }
  }
}