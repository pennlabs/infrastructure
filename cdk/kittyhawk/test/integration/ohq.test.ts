import { Construct } from 'constructs';
import { CronJob } from '../../src/cronjob';
import { Application, DjangoApplication, ReactApplication, RedisApplication } from '../../src/application'
import { chartTest } from '../utils'
import cronTime from 'cron-time-generator';

export function buildOHQChart(scope: Construct) {

  /** OHQ 
   * https://github.com/pennlabs/office-hours-queue/blob/master/k8s/values.yaml
   */
  const common = {
    image: 'pennlabs/office-hours-queue-backend',
    secret: 'office-hours-queue',
  }

  const djangoCommon = {
    ...common,
    djangoSettingsModule: 'officehoursqueue.settings.production',
    domain: 'ohq.io',
    extraEnv: [
      { name: 'REDIS_URL', value: 'redis://office-hours-queue-redis:6379' }],
  }

  new DjangoApplication(scope, 'django-asgi', {
    ...djangoCommon,
    cmd: ['/usr/local/bin/asgi-run'],
    replicas: 2,
    ingressPaths: ['/api/ws'],
  })

  new DjangoApplication(scope, 'django-wsgi', {
    ...djangoCommon,
    replicas: 4,
    ingressPaths: ['/api', '/admin', '/assets'],
  })

  new ReactApplication(scope, 'react', {
    image: common.image,
    domain: djangoCommon.domain,
    replicas: 2,
    ingressPaths: ['/'],
    portEnv: '80',
  })

  new RedisApplication(scope, 'redis', {})

  new Application(scope, 'celery', {
    ...common,
    cmd: ['celery', '-A', 'officehoursqueue', 'worker', '-lINFO'],
    extraEnv: [{ name: 'DJANGO_SETTINGS_MODULE', value: 'officehoursqueue.settings.production' },
      { name: 'REDIS_URL', value: 'redis://office-hours-queue-redis:6379' }],
  })

  new CronJob(scope, 'calculate-waits', {
    schedule: cronTime.every(5).minutes(),
    ...common,
    cmd: ['python', 'manage.py', 'calculatewaittimes'],
  });

}

test('OHQ', () => chartTest(buildOHQChart));
