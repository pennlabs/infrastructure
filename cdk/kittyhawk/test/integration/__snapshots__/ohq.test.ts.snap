// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OHQ 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-django-asgi",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-django-asgi",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-django-asgi",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-django-asgi",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 2,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-django-asgi",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-django-asgi",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "command": Array [
                "/usr/local/bin/asgi-run",
              ],
              "env": Array [
                Object {
                  "name": "REDIS_URL",
                  "value": "redis://office-hours-queue-redis:6379",
                },
                Object {
                  "name": "DJANGO_SETTINGS_MODULE",
                  "value": "officehoursqueue.settings.production",
                },
                Object {
                  "name": "DOMAIN",
                  "value": "ohq.io",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "office-hours-queue",
                  },
                },
              ],
              "image": "pennlabs/office-hours-queue-backend:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-django-asgi",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "ohq.io",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-django-asgi",
                  "servicePort": 80,
                },
                "path": "/api/ws",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "ohq.io",
          ],
          "secretName": "ohq-io-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "ohq-io",
    },
    "spec": Object {
      "dnsNames": Array [
        "ohq.io",
        "*.ohq.io",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "ohq-io-tls",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-django-wsgi",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-django-wsgi",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-django-wsgi",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-django-wsgi",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 4,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-django-wsgi",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-django-wsgi",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "REDIS_URL",
                  "value": "redis://office-hours-queue-redis:6379",
                },
                Object {
                  "name": "DJANGO_SETTINGS_MODULE",
                  "value": "officehoursqueue.settings.production",
                },
                Object {
                  "name": "DOMAIN",
                  "value": "ohq.io",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "office-hours-queue",
                  },
                },
              ],
              "image": "pennlabs/office-hours-queue-backend:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-django-wsgi",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "ohq.io",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-django-wsgi",
                  "servicePort": 80,
                },
                "path": "/api",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-django-wsgi",
                  "servicePort": 80,
                },
                "path": "/admin",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-django-wsgi",
                  "servicePort": 80,
                },
                "path": "/assets",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "ohq.io",
          ],
          "secretName": "ohq-io-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "ohq-io",
    },
    "spec": Object {
      "dnsNames": Array [
        "ohq.io",
        "*.ohq.io",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "ohq-io-tls",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-react",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-react",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-react",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-react",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 2,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-react",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-react",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "DOMAIN",
                  "value": "ohq.io",
                },
                Object {
                  "name": "PORT",
                  "value": "80",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "image": "pennlabs/office-hours-queue-backend:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-react",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "ohq.io",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-react",
                  "servicePort": 80,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "ohq.io",
          ],
          "secretName": "ohq-io-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "ohq-io",
    },
    "spec": Object {
      "dnsNames": Array [
        "ohq.io",
        "*.ohq.io",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "ohq-io-tls",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-redis",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 6379,
          "targetPort": 6379,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-redis",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-redis",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-redis",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-redis",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-redis",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "image": "redis:6.0",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 6379,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-celery",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-celery",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-celery",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-celery",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-celery",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-celery",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "command": Array [
                "celery",
                "-A",
                "officehoursqueue",
                "worker",
                "-lINFO",
              ],
              "env": Array [
                Object {
                  "name": "DJANGO_SETTINGS_MODULE",
                  "value": "officehoursqueue.settings.production",
                },
                Object {
                  "name": "REDIS_URL",
                  "value": "redis://office-hours-queue-redis:6379",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "office-hours-queue",
                  },
                },
              ],
              "image": "pennlabs/office-hours-queue-backend:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "batch/v1beta1",
    "kind": "CronJob",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-calculate-waits",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-calculate-waits",
      "namespace": "default",
    },
    "spec": Object {
      "failedJobsHistoryLimit": 1,
      "jobTemplate": Object {
        "spec": Object {
          "template": Object {
            "spec": Object {
              "containers": Array [
                Object {
                  "command": Array [
                    "python",
                    "manage.py",
                    "calculatewaittimes",
                  ],
                  "env": Array [
                    Object {
                      "name": "GIT_SHA",
                      "value": "TAG_FROM_CI",
                    },
                  ],
                  "envFrom": Array [
                    Object {
                      "secretRef": Object {
                        "name": "office-hours-queue",
                      },
                    },
                  ],
                  "image": "pennlabs/office-hours-queue-backend:TAG_FROM_CI",
                  "imagePullPolicy": "IfNotPresent",
                  "name": "worker",
                  "ports": Array [
                    Object {
                      "containerPort": 80,
                    },
                  ],
                },
              ],
              "restartPolicy": "Never",
            },
          },
        },
      },
      "schedule": "*/5 * * * *",
      "successfulJobsHistoryLimit": 1,
    },
  },
]
`;
